### PR TITLE
Allow training for DS3.2 lightning indexer

### PR DESCRIPTION
**Pull request by Rocketknight1**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48863) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48863)

This PR adds training support for the DS3.2 lightning indexer. There are a bunch of other models that also use this or similar layers (glm_moe_dsa, hyv4, etc.) Currently the indexer can't be trained on these models and is inference-only.

I'll add support for other models later, but I think it's easiest to do the initial review just for one model! Once we get the PR structure right, we can just copy that structure to other models with small model-specific tweaks.

### CODE REVIEW

File: src/transformers/models/axk2/modeling_axk2.py [hunk 1]
```diff
@@ -231,61 +228,54 @@ def forward(
...
-        The scoring logic computes:
-            index_score[b,s,t] = Σ_h (weight[b,s,h] · softmax_scale · q[b,s,h,:] · k[b,t,:])
-
-        Args:
-            hidden_states: Input hidden states `[B, S, hidden_size]`.
-            q_resid: Query residual from `q_a_layernorm(q_a_proj(x))`, shape `[B, S, q_lora_rank]`.
-            position_embeddings: `(cos, sin)` from RotaryEmbedding.
-            attention_mask: Causal mask, broadcastable to `[B, S, T]`.
-            past_key_values: Cache object containing the indexer key cache for this layer.
+        output_scores: bool = False,
```
Comment (ArthurZucker) [inline src/transformers/models/axk2/modeling_axk2.py]: no what we do is we always return the scores, see router or attention

File: src/transformers/models/deepseek_v32/modeling_deepseek_v32.py [hunk 2]
```diff
@@ -197,61 +195,54 @@ def forward(
...
-        """
-        Selects the top-k tokens per query for DeepSeek Sparse Attention (DSA).
-
-        This is the bf16 equivalent of the reference Indexer which uses `rotate_activation` (Hadamard transform)
-        and `fp8_index` (FP8 quantized scoring kernel). Since the Hadamard transform is orthogonal (dot products
-        are preserved: Hq·Hk = q·k), and FP8 quantization is a precision optimization, we skip both and compute
-        scores directly in bf16/fp32.
-
-        The scoring logic computes:
-            index_score[b,s,t] = Σ_h (weight[b,s,h] · softmax_scale · q[b,s,h,:] · k[b,t,:])
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v32/modeling_deepseek_v32.py:210]: why are we removing valuable comments?

File: src/transformers/models/deepseek_v32/modeling_deepseek_v32.py [hunk 3]
```diff
@@ -346,6 +337,40 @@ def apply_rotary_pos_emb_interleave(q, k, cos, sin, position_ids=None, unsqueeze
     return q_embed, k_embed
 
 
+def indexer_kl_loss(
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v32/modeling_deepseek_v32.py]: let's not put it in model maybe?

File: src/transformers/models/deepseek_v32/modeling_deepseek_v32.py [hunk 4]
```diff
@@ -445,15 +472,26 @@ def forward(
...
             q_resid,
             position_embeddings,
             attention_mask[:, 0, :, :],
             position_ids,  # Kept for BC
             past_key_values=past_key_values,
-        )  # [B, S, topk]
+            output_scores=output_indexer_loss,
+        )
+        indexer_loss = None
+        if output_indexer_loss:
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v32/modeling_deepseek_v32.py]: nope

Comment (ArthurZucker) [reply src/transformers/models/deepseek_v32/modeling_deepseek_v32.py]: this is never seen before haha

**Review by ArthurZucker — COMMENTED** [review]
Stopping the review, ty for taking the initiative, realized only later on that it might not be ready yet! 

Main pointer is to use the OutputRecorder, kwargs, and make it as transparent as possible. The only change that is valid is to re-organise the forward of the indexer to make it return both scores and index, but that should be it in general! 
🤗

**Comment by Rocketknight1** [comment]
It was still a draft, I'm sorry! I just marked it ready for review to check the CI. I'll take those into account before I ping the reviewers properly, though!

File: src/transformers/models/axk2/modeling_axk2.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/models/axk2/modeling_axk2.py]: Sorry, yeah! I was testing just computing the loss because returning all of the logits was really big and made training go OOM. I got it working just keeping the topk (which is all the loss needs), so it matches `aux_loss` better now

File: src/transformers/models/deepseek_v32/modeling_deepseek_v32.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/models/deepseek_v32/modeling_deepseek_v32.py:210]: This one was accidental, sorry! Fixed

File: src/transformers/models/deepseek_v32/modeling_deepseek_v32.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/models/deepseek_v32/modeling_deepseek_v32.py]: I moved it out of the main stem forward! I can move it to `loss/` instead, but other models with QSAIndexer / lightning indexer compute it slightly differently, so we might need multiple functions in there. Probably makes sense to keep it in the `modeling/modular` file

File: src/transformers/models/deepseek_v32/modeling_deepseek_v32.py (hunk 4 above)
Comment (Rocketknight1) [reply src/transformers/models/deepseek_v32/modeling_deepseek_v32.py]: Yeah, I replaced computing loss in the layers with an `OutputRecorder` approach instead!

**Comment by Rocketknight1** [comment]
@ArthurZucker should be cleaner and ready for review now! I tried to limit it just to Deepseek v3.2 - after we're happy with the structure, I'll copy it to the other models using lightning indexer / QSAIndexer. There are a few modular changes in other models, but that's to stop those models changing when we change the model they inherit from
